server/src/main/java/org/elasticsearch/index/mapper/SearchFields.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.analysis.Analyzer;
+import org.elasticsearch.index.analysis.FieldNameAnalyzer;
+import org.elasticsearch.index.analysis.IndexAnalyzers;
+import org.elasticsearch.index.query.QueryShardException;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Wraps the {@link MapperService} and exposes a subset of its functionality to the search layer.
+ * This is where runtime fields are also exposed to the search layer, hence it is important that {@link MapperService}
+ * does not leak as it exposes the ability to iterate through field mappers etc. which are not runtime fields aware.
+ */
+public class SearchFields {
+    private final MapperService mapperService;
+    private boolean allowUnmappedFields;
+    private boolean mapUnmappedFieldAsText;
+
+    public SearchFields(MapperService mapperService) {
+        this(mapperService, mapperService.getIndexSettings().isDefaultAllowUnmappedFields());
+    }
+
+    public SearchFields(SearchFields searchFields) {
+        this(searchFields.mapperService);
+    }
+
+    //this is here only for testing: it allows for a null MapperService
+    SearchFields(MapperService mapperService, boolean allowUnmappedFields) {
```

Comment:
If you have a private ctor with a static method you can name it something like that'll make it obvious that it is for testing.